lib/hurd.c

```diff
@@ -0,0 +1,381 @@
+	  snprintf(server, NAME_MAX, "%s/%04x/%02x/%02x/%01u/%s",
+		   _SERVERS_BUS_PCI, domain, bus, dev, func, entry->d_name);
+	  device_port = file_name_lookup(server, 0, 0);
+	  if (device_port == MACH_PORT_NULL)
+	    {
+	      closedir(dir);
+	      return errno;
+	    }
+
+	  d = pci_alloc_dev(a);
+	  *((mach_port_t *) d->aux) = device_port;
+	  d->bus = bus;
+	  d->dev = dev;
+	  d->func = func;
+	  pci_link_dev(a, d);
+
+	  vd = pci_read_long(d, PCI_VENDOR_ID);
+	  ht = pci_read_byte(d, PCI_HEADER_TYPE);
+
+	  d->vendor_id = vd & 0xffff;
+	  d->device_id = vd >> 16U;
+	  d->known_fields = PCI_FILL_IDENT;
+	  d->hdrtype = ht;
+	}
+    }
+
+  if (closedir(dir) < 0)
+    return errno;
+
+  return 0;
+}
+
+/* Enumerate devices */
+static void
+hurd_scan(struct pci_access *a)
+{
+  int err;
+
+  err = enum_devices(_SERVERS_BUS_PCI, a, -1, -1, -1, -1, LEVEL_DOMAIN);
+  assert(err == 0);
+}
+
+/*
+ * Read `len' bytes to `buf'.
+ *
+ * Returns error when the number of read bytes does not match `len'.
+ */
+static int
+hurd_read(struct pci_dev *d, int pos, byte * buf, int len)
+{
+  int err;
+  size_t nread;
+  char *data;
+  mach_port_t device_port;
+
+  nread = len;
+  device_port = *((mach_port_t *) d->aux);
+  if (len > 4)
+    err = !pci_generic_block_read(d, pos, buf, nread);
+  else
+    {
+      data = (char *) buf;
+      err = pci_conf_read(device_port, pos, &data, &nread, len);
+
+      if (data != (char *) buf)
+	{
+	  if (nread > (size_t) len)	/* Sanity check for bogus server.  */
+	    {
+	      vm_deallocate(mach_task_self(), (vm_address_t) data, nread);
+	      return 0;
+	    }
+
+	  memcpy(buf, data, nread);
+	  vm_deallocate(mach_task_self(), (vm_address_t) data, nread);
+	}
+    }
+  if (err)
+    return 0;
+
+  return nread == (size_t) len;
+}
+
+/*
+ * Write `len' bytes from `buf'.
+ *
+ * Returns error when the number of written bytes does not match `len'.
+ */
+static int
+hurd_write(struct pci_dev *d, int pos, byte * buf, int len)
+{
+  int err;
+  size_t nwrote;
+  mach_port_t device_port;
+
+  nwrote = len;
+  device_port = *((mach_port_t *) d->aux);
+  if (len > 4)
+    err = !pci_generic_block_write(d, pos, buf, len);
+  else
+    err = pci_conf_write(device_port, pos, (char *) buf, len, &nwrote);
+  if (err)
+    return 0;
+
+  return nwrote == (size_t) len;
+}
+
+/* Get requested info from the server */
+static int
+hurd_fill_info(struct pci_dev *d, int flags)
+{
+  int err, i;
+  struct pci_bar regions[6];
+  struct pci_xrom_bar rom;
+  size_t size;
+  char *buf;
+  mach_port_t device_port;
+
+  device_port = *((mach_port_t *) d->aux);
+
+  if (flags & PCI_FILL_IDENT)
+    {
+      d->vendor_id = pci_read_word(d, PCI_VENDOR_ID);
+      d->device_id = pci_read_word(d, PCI_DEVICE_ID);
+    }
+
+  if (flags & PCI_FILL_CLASS)
+    d->device_class = pci_read_word(d, PCI_CLASS_DEVICE);
+
+  if (flags & PCI_FILL_IRQ)
+    d->irq = pci_read_byte(d, PCI_INTERRUPT_LINE);
```

Comment:
You should let pci_generic_fill_info() handle that. Generally, back-ends implement fill_info by handling specific things and then letting the generic code handle the rest.

Reply:
Done in 5535c72